### PR TITLE
Implement container operators API function

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: docs/02.API-REFERENCE.md [hunk 1]
```diff
@@ -2322,6 +2322,89 @@ main (void)
...
+                           jerry_value_t *arguments, /**< list of arguments */
+                           uint32_t arguments_number) /**< number of arguments */
+```
+
+ - 'operation' - container operation
+ - 'this_arg' - this value
+ - 'arguments' - list of arguments
+ - 'arguments_number' - number of arguments
+ - return_value
+   - error, if argument has an error flag or operation is unsuccessful or unsupported
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:2367]: Error, on unsuccessful operation
jerry value - otherwise

File: jerry-core/api/jerry.c [hunk 2]
```diff
@@ -5110,6 +5110,90 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
   return JERRY_CONTAINER_TYPE_INVALID;
 } /* jerry_get_container_type */
 
+/**
+ * Perform container operation on the given operands (add, get, set, has, delete, size, clear).
+ *
+ * @return error - if argument has an error flag or operation is unsuccessful or unsupported
+ *         jerry value - the result of the container operation
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5117]: `result of the container operation - otherwise`

File: jerry-core/api/jerry.c [hunk 3]
```diff
@@ -5110,6 +5110,90 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+    ecma_object_t *obj_p = ecma_get_object_from_value (this_arg);
+
+    if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_CLASS)
+    {
+      uint16_t type = ((ecma_extended_object_t *) obj_p)->u.class_prop.class_id;
+      ecma_extended_object_t *map_object_p = ecma_op_container_get_object (this_arg, type);
+
+      if (map_object_p == NULL)
+      {
+        return ECMA_VALUE_ERROR;
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5138]: You should return with `ecma_create_error_reference_from_context ();`

File: jerry-core/api/jerry.c [hunk 4]
```diff
@@ -5110,6 +5110,90 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+      }
+
+      switch (operation)
+      {
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET)
+        case JERRY_CONTAINER_OP_ADD:
+        {
+          return jerry_return (ecma_op_container_set (map_object_p, arguments[0], arguments[0], type));
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5146]: You should validate the existence of `arguments[0]` before using it.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: docs/02.API-REFERENCE.md [hunk 5]
```diff
@@ -2322,6 +2335,79 @@ main (void)
...
+  runtime with the `JERRY_FEATURE_MAP, JERRY_FEATURE_SET, JERRY_FEATURE_WEAKMAP, JERRY_FEATURE_WEAKSET`
+  feature enum values.
+  see: [jerry_is_feature_enabled](#jerry_is_feature_enabled).
+- The es.next profile enables this by default.
+
+**Prototype**
+
+```c
+jerry_value_t
+jerry_contaiter_operation (jerry_container_operation_t operation,
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md:2359]: jerry_contaiter_operation -> jerry_container_operation

File: jerry-core/api/jerry.c [hunk 6]
```diff
@@ -5110,6 +5110,97 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        }
+#endif /* ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET) */
+#if ENABLED (JERRY_BUILTIN_MAP) || ENABLED (JERRY_BUILTIN_WEAKMAP)
+        case JERRY_CONTAINER_OP_GET:
+        {
+          return jerry_return (ecma_op_container_get (map_object_p, arguments[0], type));
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5159]: I would put these into a local result.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: docs/02.API-REFERENCE.md [hunk 7]
```diff
@@ -391,6 +391,19 @@ An opaque declaration of the JerryScript context structure.
 typedef struct jerry_context_t jerry_context_t;
 ```
 
+*New in version [[NEXT_RELEASE]]*.
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:394]: This seems to be at an incorrect place. It should be under the new enum type (after the description)

File: docs/02.API-REFERENCE.md [hunk 8]
```diff
@@ -2322,6 +2335,79 @@ main (void)
...
+
+**Summary**
+
+Perform container operation on the given operands (add, delete, set, etc.).
+
+*Note*:
+- Returned value must be freed with [jerry_release_value](#jerry_release_value) when it
+  is no longer needed.
+- This API function depends on a build option (`JERRY_BUILTIN_CONTAINER`) and can be checked
+  runtime with the `JERRY_FEATURE_MAP, JERRY_FEATURE_SET, JERRY_FEATURE_WEAKMAP, JERRY_FEATURE_WEAKSET`
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2350]: So... this method works on weaksets?

File: docs/02.API-REFERENCE.md [hunk 9]
```diff
@@ -391,6 +391,19 @@ An opaque declaration of the JerryScript context structure.
...
 
+*New in version [[NEXT_RELEASE]]*.
+
+## jerry_container_operation_t
+
+Enum that contains the supported container operation types
+  - JERRY_CONTAINER_OP_ADD - Set.prototype.add
+  - JERRY_CONTAINER_OP_GET - Map.prototype.get
+  - JERRY_CONTAINER_OP_SET - Map.prototype.set
+  - JERRY_CONTAINER_OP_HAS - Map.prototype.has
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:402]: Does this means that the `..HAS` only works for Map objects?

File: jerry-core/api/jerry.c [hunk 10]
```diff
@@ -5110,6 +5110,97 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        {
+          return jerry_return (ecma_op_container_has (map_object_p, arguments[0], type));
+        }
+        case JERRY_CONTAINER_OP_DELETE:
+        {
+          if (type == LIT_MAGIC_STRING_WEAKMAP_UL || type  == LIT_MAGIC_STRING_WEAKSET_UL)
+          {
+            return jerry_return (ecma_op_container_delete (map_object_p, arguments[0], type));
+          }
+          return jerry_return (ecma_op_container_delete (map_object_p, arguments[0], type));
```
Comment (galpeter) [inline jerry-core/api/jerry.c]: What is  the difference between the method call in the `if` and outside of it?

File: jerry-core/api/jerry.c [hunk 11]
```diff
@@ -5110,6 +5110,100 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+          result = jerry_return (ecma_op_container_has (map_object_p, arguments[0], type));
+        }
+        case JERRY_CONTAINER_OP_DELETE:
+        {
+          if (type == LIT_MAGIC_STRING_WEAKMAP_UL || type  == LIT_MAGIC_STRING_WEAKSET_UL)
+          {
+            result = jerry_return (ecma_op_container_delete (map_object_p, arguments[0], type));
+          }
+          result = jerry_return (ecma_op_container_delete (map_object_p, arguments[0], type));
+        }
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5181]: What is the difference between the two cases (the on inside the if and outside)?

File: jerry-core/api/jerry.c [hunk 12]
```diff
@@ -5110,6 +5110,100 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+
+  if (ecma_is_value_object (container))
+  {
+#if ENABLED (JERRY_BUILTIN_CONTAINER)
+    ecma_object_t *obj_p = ecma_get_object_from_value (container);
+
+    if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_CLASS)
+    {
+      uint16_t type = ((ecma_extended_object_t *) obj_p)->u.class_prop.class_id;
+      ecma_extended_object_t *map_object_p = ecma_op_container_get_object (container, type);
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5141]: If this works for all kind of containers, then the `map_object_p` name is misleading.

File: jerry-core/api/jerry.c [hunk 13]
```diff
@@ -5110,6 +5110,100 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+      }
+
+      jerry_value_t result;
+
+      switch (operation)
+      {
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET)
+        case JERRY_CONTAINER_OP_ADD:
+        {
+          result = jerry_return (ecma_op_container_set (map_object_p, arguments[0], arguments[0], type));
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5155]: is it really ok that there are no `break` -s in the cases?

File: jerry-core/api/jerry.c [hunk 14]
```diff
@@ -5110,6 +5110,100 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+{
+  jerry_assert_api_available ();
+
+  if ((arguments_number >= 1 && ecma_is_value_error_reference (arguments[0]))
+      || (arguments_number >= 2 && ecma_is_value_error_reference (arguments[1])))
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p)));
+  }
+
+  if (ecma_is_value_object (container))
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5133]: So if the input `container` is no an object the error message would be that the containers are disabled?

File: tests/unit-core/test-container-operators.c [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_SIZE, set, NULL, 0);
+  TEST_ASSERT (jerry_get_number_value (value) == 1);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_CLEAR, set, NULL, 0);
+  TEST_ASSERT (jerry_value_is_undefined (value));
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_SIZE, set, NULL, 0);
+  TEST_ASSERT (jerry_get_number_value (value) == 0);
+
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:110]: Missing tests:
* providing a non-container type as the input argument
* providing a container type as input object, but incorrectly specifying the argument numbers.
* returning a non trivial type (eg: objects) to test if the returned object if correctly refcounted (must call the `jerry_release_value` )

File: tests/unit-core/test-container-operators.c [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+  TEST_ASSERT (jerry_get_container_type (map) == JERRY_CONTAINER_TYPE_MAP);
+
+  jerry_value_t key_str = jerry_create_string ((jerry_char_t *) "number");
+  jerry_value_t number = jerry_create_number (10);
+  jerry_value_t args[2] = {key_str, number};
+  jerry_value_t value = jerry_container_operation (JERRY_CONTAINER_OP_SET, map, args, 2);
+  TEST_ASSERT (!jerry_value_is_error (value));
+  jerry_release_value (value);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_GET, map, &key_str, 1);
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:44]: If the doc describes that we need to call the `jerry_release_value` then we should call before assigning a new value to the `value` variable.

File: tests/unit-core/test-container-operators.c [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    return 0;
+  }
+
+  jerry_value_t map = jerry_create_container (JERRY_CONTAINER_TYPE_MAP, NULL, 0);
+  TEST_ASSERT (jerry_get_container_type (map) == JERRY_CONTAINER_TYPE_MAP);
+
+  jerry_value_t key_str = jerry_create_string ((jerry_char_t *) "number");
+  jerry_value_t number = jerry_create_number (10);
+  jerry_value_t args[2] = {key_str, number};
+  jerry_value_t value = jerry_container_operation (JERRY_CONTAINER_OP_SET, map, args, 2);
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:40]: What is the returned value in this case? (It's not an error, but what?)

File: tests/unit-core/test-container-operators.c [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+  TEST_ASSERT (!jerry_value_is_error (value));
+  jerry_release_value (value);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_GET, map, &key_str, 1);
+  TEST_ASSERT (jerry_get_number_value (value) == 10);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_HAS, map, &key_str, 1);
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:47]: Please group your tests somehow to avoid assigning multiple values in a big method for a single variable.
(for inspiration you can check the `test-api-promise.c` file)

**Review by galpeter — CHANGES_REQUESTED** [review]

**Comment by kisbg** [comment]
@galpeter  Thank you for the review!

File: docs/02.API-REFERENCE.md [hunk 19]
```diff
@@ -2322,6 +2334,79 @@ main (void)
...
+jerry_container_operation (jerry_container_operation_t operation,
+                           jerry_value_t container,
+                           jerry_value_t *arguments,
+                           uint32_t arguments_number)
+```
+
+ - 'operation' - container operation
+ - 'container' - this value
+ - 'arguments' - list of arguments
+ - 'arguments_number' - number of arguments
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2367]: AFAIK, we use backticks here (pleas check other API methods) instead of normal `'` characters.

File: docs/02.API-REFERENCE.md [hunk 20]
```diff
@@ -2322,6 +2334,79 @@ main (void)
...
+  jerry_value_t args[2] = {key_str, number};
+
+  jerry_value_t value = jerry_container_operation (JERRY_CONTAINER_OP_SET, map, args, 2);
+  jerry_release_value (value);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_SIZE, map, NULL, 0);
+  jerry_release_value (value);
+
+  value = jerry_container_operation (JERRY_CONTAINER_OP_CLEAR, map, NULL, 0);
+  jerry_release_value (value);
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2395]: please use different `value`  names for here.

File: docs/02.API-REFERENCE.md [hunk 21]
```diff
@@ -2322,6 +2334,79 @@ main (void)
...
+ - return_value
+   - Error, on unsuccessful operation
+   - jerry value - otherwise
+
+**Example**
+
+[doctest]: # ()
+
+```c
+#include "jerryscript.h"
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2377]: Please add a newline after the include.

File: jerry-core/api/jerry.c [hunk 22]
```diff
@@ -5110,6 +5110,122 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET) \
+        || ENABLED (JERRY_BUILTIN_MAP) || ENABLED (JERRY_BUILTIN_WEAKMAP)
+        case JERRY_CONTAINER_OP_HAS:
+        {
+          result = jerry_return (ecma_op_container_has (container_object_p, arguments[0], type));
+          break;
+        }
+        case JERRY_CONTAINER_OP_DELETE:
+        {
+          if (type == LIT_MAGIC_STRING_WEAKMAP_UL || type  == LIT_MAGIC_STRING_WEAKSET_UL)
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5186]: There is a double space before the second `==`

File: jerry-core/api/jerry.c [hunk 23]
```diff
@@ -5110,6 +5110,122 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+    JERRY_UNUSED (container);
+    JERRY_UNUSED (arguments);
+    JERRY_UNUSED (arguments_number);
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not a object.")));;
+  }
+  JERRY_UNUSED (operation);
+  JERRY_UNUSED (container);
+  JERRY_UNUSED (arguments);
+  JERRY_UNUSED (arguments_number);
+  return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Containers are disabled.")));;
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5226]: I would throw this error if containers are disabled regardless of input.

File: jerry-core/api/jerry.c [hunk 24]
```diff
@@ -5110,6 +5110,122 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+    }
+#endif /* ENABLED (JERRY_BUILTIN_CONTAINER) */
+  }
+  else
+  {
+    JERRY_UNUSED (operation);
+    JERRY_UNUSED (container);
+    JERRY_UNUSED (arguments);
+    JERRY_UNUSED (arguments_number);
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not a object.")));;
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5220]: If this else is here, when will the lines below it run? This `else` is only required if the containers are enabled, in every other case the normal "disabled" error can be thrown.

File: jerry-core/api/jerry.c (hunk 14 above)
Comment (zherczeg) [reply jerry-core/api/jerry.c:5133]: We could use an early return here.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: tests/unit-core/test-container-operators.c [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+  value = jerry_container_operation (JERRY_CONTAINER_OP_SET, map, args3, 2);
+  TEST_ASSERT (jerry_value_is_error (value));
+  jerry_release_value (value);
+  jerry_release_value (error_val);
+
+  jerry_release_value (set);
+  jerry_release_value (key_str);
+  jerry_release_value (number);
+  jerry_release_value (map);
+
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:140]: hmm should we also test the weakmap/weakset?

File: docs/02.API-REFERENCE.md (hunk 19 above)
Comment (kisbg) [reply docs/02.API-REFERENCE.md:2367]: Yes you are right.

File: jerry-core/api/jerry.c [hunk 26]
```diff
@@ -5281,6 +5281,117 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        default:
+        {
+          result = jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Unsupported container operation")));
+        }
+      }
+      return result;
+    }
+  }
+  JERRY_UNUSED (operation);
+  return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not a object.")));
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5385]: We could use an early return when the container is not valid. It would be less indentation.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/api/jerry.c [hunk 27]
```diff
@@ -5281,6 +5281,116 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+#if ENABLED (JERRY_BUILTIN_CONTAINER)
+  if ((arguments_number >= 1 && ecma_is_value_error_reference (arguments[0]))
+      || (arguments_number >= 2 && ecma_is_value_error_reference (arguments[1])))
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p)));
+  }
+
+  if (!ecma_is_value_object (container))
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not a object.")));
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5306]: `is not an`

Comment (kisbg) [reply jerry-core/api/jerry.c]: True!

File: docs/02.API-REFERENCE.md (hunk 8 above)
Comment (rerobika) [reply docs/02.API-REFERENCE.md:2388]: It should work on them. Also:
`JERRY_FEATURE_MAP, JERRY_FEATURE_SET, JERRY_FEATURE_WEAKMAP, JERRY_FEATURE_WEAKSET` ->
`JERRY_FEATURE_MAP`, `JERRY_FEATURE_SET`, `JERRY_FEATURE_WEAKMAP`, `JERRY_FEATURE_WEAKSET`

File: jerry-core/api/jerry.c [hunk 28]
```diff
@@ -5281,6 +5281,117 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        break;
+      }
+#endif /* ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET) || ENABLED (JERRY_BUILTIN_MAP)
+          || ENABLED (JERRY_BUILTIN_WEAKMAP) */
+      default:
+      {
+        result = jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Unsupported container operation")));
+      }
+    }
+    return result;
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5383]: use the `jerry_return` here, not in all cases.

File: jerry-core/api/jerry.c [hunk 29]
```diff
@@ -5281,6 +5281,117 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+
+        if (!jerry_value_is_error (result))
+        {
+          jerry_release_value (result);
+          result = jerry_create_boolean (true);
+        }
+        break;
+      }
+#endif /* ENABLED (JERRY_BUILTIN_MAP) || ENABLED (JERRY_BUILTIN_WEAKMAP) */
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET) \
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5349]: This equivalent to `ENABLED (JERRY_BUILTIN_CONTAINER)` which is already guarded, so it can be removed.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/api/jerry.c [hunk 30]
```diff
@@ -5281,6 +5281,113 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        break;
+      }
+      default:
+      {
+        result = jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Unsupported container operation")));
+      }
+    }
+    return jerry_return (result);
+  }
+  return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Containers are disabled.")));
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5381]: I think the error messages should be swapped
Containers are disabled <-> Container is not an object.

Also, we could do an early return (when `if (ecma_get_object_type (obj_p) != ECMA_OBJECT_TYPE_CLASS)`)

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Review by zherczeg — APPROVED** [review]
LGTM

File: jerry-core/api/jerry.c [hunk 31]
```diff
@@ -5281,6 +5281,118 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+      }
+      case JERRY_CONTAINER_OP_SET:
+      {
+        if (arguments_number < 2)
+        {
+          return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("set operation needs two arguments")));
+        }
+        result = ecma_op_container_set (container_object_p, arguments[0], arguments[1], type);
+
+        if (!jerry_value_is_error (result))
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5345]: This should be `ECMA_IS_VALUE_ERROR`

File: jerry-core/api/jerry.c [hunk 32]
```diff
@@ -5281,6 +5281,118 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+      return ecma_create_error_reference_from_context ();
+    }
+
+    jerry_value_t result;
+    switch (operation)
+    {
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET)
+      case JERRY_CONTAINER_OP_ADD:
+      {
+        result = jerry_return (ecma_op_container_set (container_object_p, arguments[0], arguments[0], type));
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5327]: `jerry_return` is not needed.

File: jerry-core/api/jerry.c [hunk 33]
```diff
@@ -5281,6 +5281,118 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        break;
+      }
+#endif /* ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET) */
+#if ENABLED (JERRY_BUILTIN_MAP) || ENABLED (JERRY_BUILTIN_WEAKMAP)
+      case JERRY_CONTAINER_OP_GET:
+      {
+        result = jerry_return (ecma_op_container_get (container_object_p, arguments[0], type));
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5334]: ditto.

File: jerry-core/api/jerry.c [hunk 34]
```diff
@@ -5281,6 +5281,114 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+jerry_value_t
+jerry_container_operation (jerry_container_operation_t operation, /**< container operation */
+                           jerry_value_t container, /**< container */
+                           jerry_value_t *arguments, /**< list of arguments */
+                           uint32_t arguments_number) /**< number of arguments */
+{
+  jerry_assert_api_available ();
+#if ENABLED (JERRY_BUILTIN_CONTAINER)
+  if ((arguments_number >= 1 && ecma_is_value_error_reference (arguments[0]))
+      || (arguments_number >= 2 && ecma_is_value_error_reference (arguments[1])))
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/api/jerry.c:5299]: I think all this validation must be performed per operation not in general.

File: jerry-core/api/jerry.c (hunk 32 above)
Comment (rerobika) [reply jerry-core/api/jerry.c:5327]: accessing `arguments[0]` is unsafe currently. please validate `argc` before (same in all other operations below),

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/api/jerry.c [hunk 35]
```diff
@@ -5281,6 +5281,133 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (container);
+
+  if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_CLASS)
+  {
+    uint16_t type = ((ecma_extended_object_t *) obj_p)->u.class_prop.class_id;
+    ecma_extended_object_t *container_object_p = ecma_op_container_get_object (container, type);
+
+    if (container_object_p == NULL)
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5310]: Reverse the validation logic.

If (non object) return error;
if (non type_class) return error;
if (not container id) return error;

File: jerry-core/api/jerry.c [hunk 36]
```diff
@@ -5281,6 +5281,133 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+      return ecma_create_error_reference_from_context ();
+    }
+
+    jerry_value_t result;
+    switch (operation)
+    {
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET)
+      case JERRY_CONTAINER_OP_ADD:
+      {
+        if (arguments_number < 1 || (arguments_number >= 1 && ecma_is_value_error_reference (arguments[0])))
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5321]: This condition is superfluous. 
simply: `if (arguments_number < 1 || ecma_is_value_error_reference (arguments[0]))`

Same as all other cases below.

File: jerry-core/api/jerry.c [hunk 37]
```diff
@@ -5281,6 +5281,133 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+        {
+          return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p)));
+        }
+
+        result = ecma_op_container_get (container_object_p, arguments[0], type);
+        break;
+      }
+      case JERRY_CONTAINER_OP_SET:
+      {
+        if (arguments_number < 2 || (arguments_number >= 1 && ecma_is_value_error_reference (arguments[0]))
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5343]: `if (arguments_number < 2 || ecma_is_value_error_reference (arguments[0]) || || ecma_is_value_error_reference (arguments[1]))`

File: jerry-core/api/jerry.c [hunk 38]
```diff
@@ -5281,6 +5281,133 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+            || (arguments_number >= 2 && ecma_is_value_error_reference (arguments[1])))
+        {
+          return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p)));
+        }
+        result = ecma_op_container_set (container_object_p, arguments[0], arguments[1], type);
+
+        if (!ECMA_IS_VALUE_ERROR (result))
+        {
+          jerry_release_value (result);
+          result = jerry_create_boolean (true);
```
Comment (rerobika) [inline jerry-core/api/jerry.c:5353]: `result = ECMA_VALUE_TRUE;`

File: jerry-core/api/jerry.c (hunk 30 above)
Comment (rerobika) [reply jerry-core/api/jerry.c:5401]: This request is still valid. Please recheck the error messages.

**Review by rerobika — CHANGES_REQUESTED** [review]

**Comment by kisbg** [comment]
@rerobika  Thank your the review! i updated the pull request based on your review.

File: jerry-core/api/jerry.c [hunk 39]
```diff
@@ -5815,6 +5815,135 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+    return ecma_create_error_reference_from_context ();
+  }
+
+  jerry_value_t result;
+  switch (operation)
+  {
+#if ENABLED (JERRY_BUILTIN_SET) || ENABLED (JERRY_BUILTIN_WEAKSET)
+    case JERRY_CONTAINER_OP_ADD:
+    {
+      if (arguments_number < 1 || (arguments_number >= 1 && ecma_is_value_error_reference (arguments[0])))
```
Comment (galpeter) [inline jerry-core/api/jerry.c:5857]: As previously mentioned the `arguments_number>=1` part is not needed if the `arguments_number<1` is values the arg numbers are by default greater than 1.

File: docs/02.API-REFERENCE.md [hunk 40]
```diff
@@ -2412,6 +2424,81 @@ main (void)
...
+*Note*:
+- Returned value must be freed with [jerry_release_value](#jerry_release_value) when it
+  is no longer needed.
+- This API function depends on a build option (`JERRY_BUILTIN_CONTAINER`) and can be checked
+  runtime with the `JERRY_FEATURE_MAP` , `JERRY_FEATURE_SET` , `JERRY_FEATURE_WEAKMAP` , `JERRY_FEATURE_WEAKSET`
+  feature enum values.
+  see: [jerry_is_feature_enabled](#jerry_is_feature_enabled).
+- The es.next profile enables this by default.
+
+*New in version [[NEXT_RELEASE]]*.
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2442]: This "new version" is not at the correct place, It should be just before the example "block". (Just noticed in other api descriptions this is also at an incorrect place, but that should be a different PR fix)

File: tests/unit-core/test-container-operators.c [hunk 41]
```diff
@@ -0,0 +1,182 @@
...
+  // container is not a object
+  jerry_value_t empty_val = jerry_create_undefined ();
+  result = jerry_container_operation (JERRY_CONTAINER_OP_SET, empty_val, args, 2);
+  TEST_ASSERT (jerry_value_is_error (result));
+  jerry_release_value (result);
+
+  // arguments is a error
+  const char * const error_message_p = "Random error.";
+  jerry_value_t error_val = jerry_create_error (JERRY_ERROR_RANGE, (const jerry_char_t *) error_message_p);
+  jerry_value_t args3[2] = { error_val, error_val};
```
Comment (galpeter) [inline tests/unit-core/test-container-operators.c:170]: Missing space before the closing `}`.

File: docs/02.API-REFERENCE.md [hunk 42]
```diff
@@ -2412,6 +2424,81 @@ main (void)
...
+jerry_value_t
+jerry_container_operation (jerry_container_operation_t operation,
+                           jerry_value_t container,
+                           jerry_value_t *arguments,
+                           uint32_t arguments_number)
+```
+
+ - `operation` - container operation
+ - `container` - this value
+ - `arguments` - list of arguments
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:2456]: `list` -> `array`

File: docs/02.API-REFERENCE.md [hunk 43]
```diff
@@ -430,6 +430,18 @@ typedef struct jerry_context_t jerry_context_t;
...
+## jerry_container_operation_t
+
+Enum that contains the supported container operation types
+  - JERRY_CONTAINER_OP_ADD - Set/WeakSet add operation
+  - JERRY_CONTAINER_OP_GET - Map/WeakMap get operation
+  - JERRY_CONTAINER_OP_SET - Map/WeakMap set operation
+  - JERRY_CONTAINER_OP_HAS - Set/WeakSet/Map/WeakMap has operation
+  - JERRY_CONTAINER_OP_DELETE - Set/WeakSet/Map/WeakMap delete operation
+  - JERRY_CONTAINER_OP_SIZE - Set/WeakSet/Map/WeakMap size operation
+  - JERRY_CONTAINER_OP_CLEAR - Set/Map clear operation
```
Comment (galpeter) [inline docs/02.API-REFERENCE.md:442]: We should somehow/somewhere describe that for each operation what should be arguments and arguments number. Maybe at the method's description.
